Use of external Actor System in the library

When I am trying to create a TwitterRestClient it makes its own ActorSystem and I have already made one actor system for my app. Is there a way in which the internal actor system can be override i.e it will use the external actor system that is provided if not then it can create one for itself.

I am trying to build an application where I am building my own Actor System,(pretty basic Scenario).
I run this app on my laptop it works ok ! But When I deploy it on Heroku , within seconds it crashes with this error:
`2018-01-07T15:44:08.192109+00:00 app[web.1]: Uncaught error from thread [twitter4s-rest-bea6326a-21bf-41ec-a239-2de7c58431ab-akka.actor.default-dispatcher-4]: unable to create new native thread, Uncaught error from thread [twitter4s-rest-bea6326a-21bf-41ec-a239-2de7c58431ab-akka.actor.default-dispatcher-2]: unable to create new native thread, shutting down JVM since 'akka.jvm-exit-on-fatal-error' is enabled for for ActorSystem[twitter4s-rest-bea6326a-21bf-41ec-a239-2de7c58431ab]

2018-01-07T15:44:08.220463+00:00 app[web.1]: Uncaught error from thread [twitter4s-rest-85e32954-8c17-4b50-8305-0834cecf21e4-akka.actor.default-dispatcher-6]: unable to create new native thread, shutting down JVM since 'akka.jvm-exit-on-fatal-error' is enabled for for ActorSystem[Uncaught error from thread [twitter4s-rest-dfcac4fe-c31f-49a3-844c-6e7702007ad6-akka.actor.default-dispatcher-3]: unable to create new native thread, shutting down JVM since 'akka.jvm-exit-on-fatal-error' is enabled forUncaught error from thread [twitter4s-rest-6cc560ee-a047-4891-b68a-d265d4e2e9fd-akka.actor.default-dispatcher-3 for ActorSystem[twitter4s-rest-dfcac4fe-c31f-49a3-844c-6e7702007ad6]

2018-01-07T15:44:08.220648+00:00 app[web.1]: java.lang.OutOfMemoryError: unable to create new native thread

2018-01-07T15:44:08.220843+00:00 app[web.1]: 	at java.lang.Thread.start0(Native Method)

2018-01-07T15:44:08.220968+00:00 app[web.1]: 	at java.lang.Thread.start(Thread.java:717)

2018-01-07T15:44:08.221043+00:00 app[web.1]: 	at java.util.concurrent.ForkJoinPool.createWorker(ForkJoinPool.java:1486)

2018-01-07T15:44:08.221500+00:00 app[web.1]: 	at java.util.concurrent.ForkJoinPool.tryAddWorker(ForkJoinPool.java:1517)

2018-01-07T15:44:08.221582+00:00 app[web.1]: 	at java.util.concurrent.ForkJoinPool.signalWork(ForkJoinPool.java:1634)

2018-01-07T15:44:08.221871+00:00 app[web.1]: 	at java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1733)

2018-01-07T15:44:08.221941+00:00 app[web.1]: 	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1691)

2018-01-07T15:44:08.222137+00:00 app[web.1]: 	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)

2018-01-07T15:44:08.222846+00:00 app[web.1]: Uncaught error from thread [twitter4s-rest-85e32954-8c17-4b50-8305-0834cecf21e4-akka.actor.default-dispatcher-2]: unable to create new native thread, shutting down JVM since 'akka.jvm-exit-on-fatal-error' is enabled for for ActorSystem[twitter4s-rest-85e32954-8c17-4b50-8305-0834cecf21e4]

2018-01-07T15:44:08.222952+00:00 app[web.1]: java.lang.OutOfMemoryError: unable to create new native thread

2018-01-07T15:44:08.223120+00:00 app[web.1]: 	at java.lang.Thread.start0(Native Method)

2018-01-07T15:44:08.223197+00:00 app[web.1]: 	at java.lang.Thread.start(Thread.java:717)

2018-01-07T15:44:08.223273+00:00 app[web.1]: 	at java.util.concurrent.ForkJoinPool.createWorker(ForkJoinPool.java:1486)

2018-01-07T15:44:08.223369+00:00 app[web.1]: 	at java.util.concurrent.ForkJoinPool.tryAddWorker(ForkJoinPool.java:1517)

2018-01-07T15:44:08.223447+00:00 app[web.1]: 	at java.util.concurrent.ForkJoinPool.signalWork(ForkJoinPool.java:1634)

2018-01-07T15:44:08.223523+00:00 app[web.1]: 	at java.util.concurrent.ForkJoinPool.externalPush(ForkJoinPool.java:2414)

2018-01-07T15:44:08.223625+00:00 app[web.1]: 	at java.util.concurrent.ForkJoinPool.execute(ForkJoinPool.java:2630)

2018-01-07T15:44:08.223711+00:00 app[web.1]: 	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinPool.execute(AbstractDispatcher.scala:401)

2018-01-07T15:44:08.223791+00:00 app[web.1]: 	at akka.dispatch.ExecutorServiceDelegate.execute(ThreadPoolBuilder.scala:211)

2018-01-07T15:44:08.223863+00:00 app[web.1]: 	at akka.dispatch.ExecutorServiceDelegate.execute$(ThreadPoolBuilder.scala:211)

2018-01-07T15:44:08.223946+00:00 app[web.1]: 	at akka.dispatch.Dispatcher$LazyExecutorServiceDelegate.execute(Dispatcher.scala:39)

2018-01-07T15:44:08.224020+00:00 app[web.1]: 	at akka.dispatch.Dispatcher.registerForExecution(Dispatcher.scala:115)

2018-01-07T15:44:08.224094+00:00 app[web.1]: 	at akka.dispatch.MessageDispatcher.attach(AbstractDispatcher.scala:133)

2018-01-07T15:44:08.224170+00:00 app[web.1]: 	at akka.actor.dungeon.Dispatch.start(Dispatch.scala:100)

2018-01-07T15:44:08.224246+00:00 app[web.1]: 	at akka.actor.dungeon.Dispatch.start$(Dispatch.scala:98)

2018-01-07T15:44:08.224329+00:00 app[web.1]: 	at akka.actor.ActorCell.start(ActorCell.scala:374)

2018-01-07T15:44:08.224402+00:00 app[web.1]: 	at akka.actor.ActorCell.start(ActorCell.scala:374)

2018-01-07T15:44:08.224477+00:00 app[web.1]: 	at akka.actor.RepointableActorRef.point(RepointableActorRef.scala:107)

2018-01-07T15:44:08.224558+00:00 app[web.1]: 	at akka.actor.ActorCell.handleSupervise(ActorCell.scala:625)

2018-01-07T15:44:08.224644+00:00 app[web.1]: 	at akka.actor.ActorCell.supervise(ActorCell.scala:617)

2018-01-07T15:44:08.224718+00:00 app[web.1]: 	at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:468)

2018-01-07T15:44:08.224802+00:00 app[web.1]: 	at akka.actor.ActorCell.systemInvoke(ActorCell.scala:483)

2018-01-07T15:44:08.231534+00:00 app[web.1]: 	at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:282)[ERROR] [SECURITY][01/07/2018 15:44:08.229] [twitter4s-rest-dfcac4fe-c31f-49a3-844c-6e7702007ad6-akka.actor.default-dispatcher-3] [akka.actor.ActorSystemImpl(twitter4s-rest-dfcac4fe-c31f-49a3-844c-6e7702007ad6)] Uncaught error from thread [twitter4s-rest-dfcac4fe-c31f-49a3-844c-6e7702007ad6-akka.actor.default-dispatcher-3]: unable to create new native thread, shutting down JVM since 'akka.jvm-exit-on-fatal-error' is enabled for ActorSystem[twitter4s-rest-dfcac4fe-c31f-49a3-844c-6e7702007ad6]

2018-01-07T15:44:08.231538+00:00 app[web.1]: java.lang.OutOfMemoryError: unable to create new native thread

2018-01-07T15:44:08.231539+00:00 app[web.1]: 	at java.lang.Thread.start0(Native Method)

2018-01-07T15:44:08.231540+00:00 app[web.1]: 	at java.lang.Thread.start(Thread.java:717)

2018-01-07T15:44:08.231542+00:00 app[web.1]: 	at java.util.concurrent.ForkJoinPool.createWorker(ForkJoinPool.java:1486)

2018-01-07T15:44:08.231543+00:00 app[web.1]: 	at java.util.concurrent.ForkJoinPool.tryAddWorker(ForkJoinPool.java:1517)

2018-01-07T15:44:08.231544+00:00 app[web.1]: 	at java.util.concurrent.ForkJoinPool.signalWork(ForkJoinPool.java:1634)

2018-01-07T15:44:08.231545+00:00 app[web.1]: 	at java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1733)

2018-01-07T15:44:08.231546+00:00 app[web.1]: 	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1691)

2018-01-07T15:44:08.231548+00:00 app[web.1]: 	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)

2018-01-07T15:44:08.231553+00:00 app[web.1]: 

2018-01-07T15:44:08.241659+00:00 app[web.1]: 

2018-01-07T15:44:08.241716+00:00 app[web.1]: 	at akka.dispatch.Mailbox.run(Mailbox.scala:223)

2018-01-07T15:44:08.241823+00:00 app[web.1]: 	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)

2018-01-07T15:44:08.241893+00:00 app[web.1]: 	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)

2018-01-07T15:44:08.241950+00:00 app[web.1]: 	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)

2018-01-07T15:44:08.242096+00:00 app[web.1]: 	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)

2018-01-07T15:44:08.242101+00:00 app[web.1]: 	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)

2018-01-07T15:44:08.242666+00:00 app[web.1]: shutting down JVM since 'akka.jvm-exit-on-fatal-error' is enabled for for ActorSystem[twitter4s-rest-bea6326a-21bf-41ec-a239-2de7c58431ab]

2018-01-07T15:44:08.242745+00:00 app[web.1]: java.lang.OutOfMemoryError: unable to create new native thread`

Hence I investigated and found out that multiple actor system is being made and twitter4s tries to create more actors and fails and crashes the application. Is there any way in which I can overcome this ? 
Or maybe use your actor system somehow ?

--- a/src/main/scala/com/danielasfregola/twitter4s/TwitterAuthenticationClient.scala
+++ b/src/main/scala/com/danielasfregola/twitter4s/TwitterAuthenticationClient.scala
@@ -1,13 +1,20 @@
 package com.danielasfregola.twitter4s
 
+import akka.actor.ActorSystem
 import com.danielasfregola.twitter4s.entities.ConsumerToken
 import com.danielasfregola.twitter4s.http.clients.authentication.AuthenticationClient
 import com.danielasfregola.twitter4s.http.clients.authentication.oauth.TwitterOAuthClient
 import com.danielasfregola.twitter4s.util.Configurations._
+import com.danielasfregola.twitter4s.util.SystemShutdown
 
 /** Represents the functionalities offered by the Twitter REST API for authentication
   */
-class TwitterAuthenticationClient(val consumerToken: ConsumerToken) extends TwitterAuthClients {
+class TwitterAuthenticationClient(val consumerToken: ConsumerToken)(implicit _system: ActorSystem = ActorSystem(
+                                                                      "twitter4s-auth"))
+    extends TwitterAuthClients
+    with SystemShutdown {
+
+  protected val system = _system
 
   protected val authenticationClient = new AuthenticationClient(consumerToken)
 
@@ -24,4 +31,12 @@ object TwitterAuthenticationClient {
 
   def apply(consumerToken: ConsumerToken): TwitterAuthenticationClient =
     new TwitterAuthenticationClient(consumerToken)
+
+  def withActorSystem(system: ActorSystem): TwitterAuthenticationClient = {
+    val consumerToken = ConsumerToken(key = consumerTokenKey, secret = consumerTokenSecret)
+    withActorSystem(consumerToken)(system)
+  }
+
+  def withActorSystem(consumerToken: ConsumerToken)(system: ActorSystem): TwitterAuthenticationClient =
+    new TwitterAuthenticationClient(consumerToken)(system)
 }

--- a/src/main/scala/com/danielasfregola/twitter4s/TwitterRestClient.scala
+++ b/src/main/scala/com/danielasfregola/twitter4s/TwitterRestClient.scala
@@ -1,5 +1,6 @@
 package com.danielasfregola.twitter4s
 
+import akka.actor.ActorSystem
 import com.danielasfregola.twitter4s.entities.{AccessToken, ConsumerToken}
 import com.danielasfregola.twitter4s.http.clients.rest.RestClient
 import com.danielasfregola.twitter4s.http.clients.rest.account.TwitterAccountClient
@@ -22,10 +23,16 @@ import com.danielasfregola.twitter4s.http.clients.rest.suggestions.TwitterSugges
 import com.danielasfregola.twitter4s.http.clients.rest.trends.TwitterTrendClient
 import com.danielasfregola.twitter4s.http.clients.rest.users.TwitterUserClient
 import com.danielasfregola.twitter4s.util.Configurations._
+import com.danielasfregola.twitter4s.util.SystemShutdown
 
 /** Represents the functionalities offered by the Twitter REST API
   */
-class TwitterRestClient(val consumerToken: ConsumerToken, val accessToken: AccessToken) extends RestClients {
+class TwitterRestClient(val consumerToken: ConsumerToken, val accessToken: AccessToken)(implicit _system: ActorSystem =
+                                                                                          ActorSystem("twitter4s-rest"))
+    extends RestClients
+    with SystemShutdown {
+
+  protected val system = _system
 
   protected val restClient = new RestClient(consumerToken, accessToken)
 
@@ -62,4 +69,14 @@ object TwitterRestClient {
 
   def apply(consumerToken: ConsumerToken, accessToken: AccessToken): TwitterRestClient =
     new TwitterRestClient(consumerToken, accessToken)
+
+  def withActorSystem(system: ActorSystem): TwitterRestClient = {
+    val consumerToken = ConsumerToken(key = consumerTokenKey, secret = consumerTokenSecret)
+    val accessToken = AccessToken(key = accessTokenKey, secret = accessTokenSecret)
+    withActorSystem(consumerToken, accessToken)(system)
+  }
+
+  def withActorSystem(consumerToken: ConsumerToken, accessToken: AccessToken)(system: ActorSystem): TwitterRestClient =
+    new TwitterRestClient(consumerToken, accessToken)(system)
+
 }

--- a/src/main/scala/com/danielasfregola/twitter4s/TwitterStreamingClient.scala
+++ b/src/main/scala/com/danielasfregola/twitter4s/TwitterStreamingClient.scala
@@ -1,15 +1,22 @@
 package com.danielasfregola.twitter4s
 
+import akka.actor.ActorSystem
 import com.danielasfregola.twitter4s.entities.{AccessToken, ConsumerToken}
 import com.danielasfregola.twitter4s.http.clients.streaming.StreamingClient
 import com.danielasfregola.twitter4s.http.clients.streaming.sites.TwitterSiteClient
 import com.danielasfregola.twitter4s.http.clients.streaming.statuses.TwitterStatusClient
 import com.danielasfregola.twitter4s.http.clients.streaming.users.TwitterUserClient
 import com.danielasfregola.twitter4s.util.Configurations._
+import com.danielasfregola.twitter4s.util.SystemShutdown
 
 /** Represents the functionalities offered by the Twitter Streaming API
   */
-class TwitterStreamingClient(val consumerToken: ConsumerToken, val accessToken: AccessToken) extends StreamingClients {
+class TwitterStreamingClient(val consumerToken: ConsumerToken, val accessToken: AccessToken)(
+    implicit _system: ActorSystem = ActorSystem("twitter4s-streaming"))
+    extends StreamingClients
+    with SystemShutdown {
+
+  protected val system = _system
 
   protected val streamingClient = new StreamingClient(consumerToken, accessToken)
 
@@ -27,4 +34,14 @@ object TwitterStreamingClient {
 
   def apply(consumerToken: ConsumerToken, accessToken: AccessToken): TwitterStreamingClient =
     new TwitterStreamingClient(consumerToken, accessToken)
+
+  def withActorSystem(system: ActorSystem): TwitterStreamingClient = {
+    val consumerToken = ConsumerToken(key = consumerTokenKey, secret = consumerTokenSecret)
+    val accessToken = AccessToken(key = accessTokenKey, secret = accessTokenSecret)
+    withActorSystem(consumerToken, accessToken)(system)
+  }
+
+  def withActorSystem(consumerToken: ConsumerToken, accessToken: AccessToken)(
+      system: ActorSystem): TwitterStreamingClient =
+    new TwitterStreamingClient(consumerToken, accessToken)(system)
 }

--- a/src/main/scala/com/danielasfregola/twitter4s/http/clients/authentication/AuthenticationClient.scala
+++ b/src/main/scala/com/danielasfregola/twitter4s/http/clients/authentication/AuthenticationClient.scala
@@ -1,7 +1,5 @@
 package com.danielasfregola.twitter4s.http.clients.authentication
 
-import java.util.UUID
-
 import akka.actor.ActorSystem
 import akka.http.scaladsl.model.HttpRequest
 import akka.stream.{ActorMaterializer, Materializer}
@@ -12,25 +10,22 @@ import com.danielasfregola.twitter4s.http.serializers.{FormSupport, FromMap}
 
 import scala.concurrent.Future
 
-private[twitter4s] class AuthenticationClient(val consumerToken: ConsumerToken) extends Client {
+private[twitter4s] class AuthenticationClient(val consumerToken: ConsumerToken)(implicit val system: ActorSystem)
+    extends Client {
 
   lazy val oauthProvider = new OAuth1Provider(consumerToken, None)
 
   private[twitter4s] implicit class RichRestHttpRequest(val request: HttpRequest) {
+    implicit val materializer = ActorMaterializer()
+    implicit val ec = materializer.executionContext
 
     def respondAs[T: Manifest](implicit fromMap: FromMap[T]): Future[T] = respondAs[T](None)
 
-    def respondAs[T: Manifest](callback: Option[String])(implicit fromMap: FromMap[T]): Future[T] = {
-      implicit val system = ActorSystem(s"twitter4s-authentication-${UUID.randomUUID}")
-      implicit val materializer = ActorMaterializer()
-      implicit val ec = materializer.executionContext
-      val response = for {
+    def respondAs[T: Manifest](callback: Option[String])(implicit fromMap: FromMap[T]): Future[T] =
+      for {
         requestWithAuth <- withOAuthHeader(callback)(materializer)(request)
         t <- sendReceiveAs[T](requestWithAuth)
       } yield t
-      response.onComplete(_ => system.terminate)
-      response
-    }
   }
 
   def sendReceiveAs[T: Manifest](httpRequest: HttpRequest)(implicit system: ActorSystem,

--- a/src/main/scala/com/danielasfregola/twitter4s/http/clients/rest/RestClient.scala
+++ b/src/main/scala/com/danielasfregola/twitter4s/http/clients/rest/RestClient.scala
@@ -1,7 +1,5 @@
 package com.danielasfregola.twitter4s.http.clients.rest
 
-import java.util.UUID
-
 import akka.actor.ActorSystem
 import akka.http.scaladsl.model.{HttpRequest, HttpResponse}
 import akka.http.scaladsl.unmarshalling.Unmarshal
@@ -12,47 +10,34 @@ import com.danielasfregola.twitter4s.http.oauth.OAuth1Provider
 
 import scala.concurrent.Future
 
-private[twitter4s] class RestClient(val consumerToken: ConsumerToken, val accessToken: AccessToken) extends Client {
+private[twitter4s] class RestClient(val consumerToken: ConsumerToken, val accessToken: AccessToken)(
+    implicit val system: ActorSystem)
+    extends Client {
 
   lazy val oauthProvider = new OAuth1Provider(consumerToken, Some(accessToken))
 
   private[twitter4s] implicit class RichRestHttpRequest(val request: HttpRequest) {
 
-    def respondAs[T: Manifest]: Future[T] = {
-      implicit val system = ActorSystem(s"twitter4s-rest-${UUID.randomUUID}")
-      implicit val materializer = ActorMaterializer()
-      implicit val ec = materializer.executionContext
-      val response = for {
+    implicit val materializer = ActorMaterializer()
+    implicit val ec = materializer.executionContext
+
+    def respondAs[T: Manifest]: Future[T] =
+      for {
         requestWithAuth <- withOAuthHeader(None)(materializer)(request)
         t <- sendReceiveAs[T](requestWithAuth)
       } yield t
-      response.onComplete(_ => system.terminate)
-      response
-    }
 
-    def respondAsRated[T: Manifest]: Future[RatedData[T]] = {
-      implicit val system = ActorSystem(s"twitter4s-rest-${UUID.randomUUID}")
-      implicit val materializer = ActorMaterializer()
-      implicit val ec = materializer.executionContext
-      val response = for {
+    def respondAsRated[T: Manifest]: Future[RatedData[T]] =
+      for {
         requestWithAuth <- withOAuthHeader(None)(materializer)(request)
         t <- sendReceiveAsRated[T](requestWithAuth)
       } yield t
-      response.onComplete(_ => system.terminate)
-      response
-    }
 
-    def sendAsFormData: Future[Unit] = {
-      implicit val system = ActorSystem(s"twitter4s-rest-${UUID.randomUUID}")
-      implicit val materializer = ActorMaterializer()
-      implicit val ec = materializer.executionContext
-      val response = for {
+    def sendAsFormData: Future[Unit] =
+      for {
         requestWithAuth <- withSimpleOAuthHeader(None)(materializer)(request)
         _ <- sendIgnoreResponse(requestWithAuth)
       } yield ()
-      response.onComplete(_ => system.terminate)
-      response
-    }
   }
 
   def sendIgnoreResponse(httpRequest: HttpRequest)(implicit system: ActorSystem,

--- a/src/main/scala/com/danielasfregola/twitter4s/http/clients/rest/media/TwitterMediaClient.scala
+++ b/src/main/scala/com/danielasfregola/twitter4s/http/clients/rest/media/TwitterMediaClient.scala
@@ -3,9 +3,7 @@ package clients.rest.media
 
 import java.io.{File, FileInputStream, InputStream}
 import java.net.URLConnection
-import java.util.UUID
 
-import akka.actor.ActorSystem
 import akka.http.scaladsl.model.Multipart._
 import akka.http.scaladsl.model.{ContentTypes, MediaType}
 import akka.stream.scaladsl.Source
@@ -98,16 +96,13 @@ trait TwitterMediaClient {
       filename.getOrElse(s"twitter4s-$mediaId.$extension")
     }
 
-    val system = ActorSystem(s"twitter4s-media-${UUID.randomUUID}")
-    implicit val ec = system.dispatcher
+    implicit val ec: ExecutionContext = restClient.system.dispatcher
 
-    val response = for {
+    for {
       details <- initMedia(size, media_type, additional_owners)
       uploads <- appendMedia(details.media_id, inputStream, filenameBuilder(details.media_id))
       finalize <- finalizeMedia(details.media_id)
     } yield finalize
-    response.onComplete(_ => system.terminate)
-    response
   }
 
   private def initMedia(size: Long, media_type: String, additional_owners: Seq[Long]): Future[MediaDetails] = {

--- a/src/main/scala/com/danielasfregola/twitter4s/http/clients/streaming/StreamingClient.scala
+++ b/src/main/scala/com/danielasfregola/twitter4s/http/clients/streaming/StreamingClient.scala
@@ -17,7 +17,8 @@ import org.json4s.native.Serialization
 import scala.concurrent.Future
 import scala.util.{Failure, Success, Try}
 
-private[twitter4s] class StreamingClient(val consumerToken: ConsumerToken, val accessToken: AccessToken)
+private[twitter4s] class StreamingClient(val consumerToken: ConsumerToken, val accessToken: AccessToken)(
+    implicit val system: ActorSystem)
     extends OAuthClient {
 
   val withLogRequest = true
@@ -29,15 +30,14 @@ private[twitter4s] class StreamingClient(val consumerToken: ConsumerToken, val a
 
   private[twitter4s] implicit class RichStreamingHttpRequest(val request: HttpRequest) {
 
-    def processStream[T <: StreamingMessage: Manifest](f: PartialFunction[T, Unit]): Future[TwitterStream] = {
-      implicit val system = ActorSystem(s"twitter4s-streaming-${UUID.randomUUID}")
-      implicit val materializer = ActorMaterializer()
-      implicit val ec = materializer.executionContext
+    implicit val materializer = ActorMaterializer()
+    implicit val ec = materializer.executionContext
+
+    def processStream[T <: StreamingMessage: Manifest](f: PartialFunction[T, Unit]): Future[TwitterStream] =
       for {
         requestWithAuth <- withOAuthHeader(None)(materializer)(request)
         killSwitch <- processStreamRequest(requestWithAuth)(f)
       } yield TwitterStream(consumerToken, accessToken)(killSwitch, requestWithAuth, system)
-    }
   }
 
   protected def processStreamRequest[T <: StreamingMessage: Manifest](request: HttpRequest)(

--- a/src/main/scala/com/danielasfregola/twitter4s/http/clients/streaming/TwitterStream.scala
+++ b/src/main/scala/com/danielasfregola/twitter4s/http/clients/streaming/TwitterStream.scala
@@ -7,7 +7,7 @@ import com.danielasfregola.twitter4s.StreamingClients
 import com.danielasfregola.twitter4s.entities.{AccessToken, ConsumerToken}
 import com.typesafe.scalalogging.LazyLogging
 
-import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.Future
 
 /** Represents a twitter stream operation. It can be used to close the stream on demand
   * or to replace the current stream with another twitter stream.
@@ -19,9 +19,7 @@ final case class TwitterStream(consumerToken: ConsumerToken, accessToken: Access
     extends StreamingClients
     with LazyLogging {
 
-  implicit private val ec: ExecutionContext = system.dispatcher
-
-  protected val streamingClient = new StreamingClient(consumerToken, accessToken) {
+  protected val streamingClient = new StreamingClient(consumerToken, accessToken)(system) {
     override def preProcessing(): Unit = {
       super.preProcessing()
       close()
@@ -32,9 +30,8 @@ final case class TwitterStream(consumerToken: ConsumerToken, accessToken: Access
     *
     * @return : Future that will be completed with Unit once the stream has been terminated.
     */
-  def close(): Future[Unit] = {
+  def close(): Future[Unit] = Future.successful {
     logger.info(s"${request.method.value} ${request.uri}: closing streaming")
     killSwitch.shutdown
-    system.terminate.map(_ => (): Unit)
   }
 }

--- a/src/main/scala/com/danielasfregola/twitter4s/util/SystemShutdown.scala
+++ b/src/main/scala/com/danielasfregola/twitter4s/util/SystemShutdown.scala
@@ -1,0 +1,17 @@
+package com.danielasfregola.twitter4s.util
+
+import akka.actor.ActorSystem
+
+import scala.concurrent.Future
+
+trait SystemShutdown {
+
+  protected def system: ActorSystem
+
+  /** Shutdown of the actor system associated to the client
+    */
+  def shutdown(): Future[Unit] = {
+    implicit val ec = system.dispatcher
+    system.terminate.map(_ => (): Unit)
+  }
+}

